File: frontend/src/objects/ui-customization/template-category.ts
import { z } from "zod";

export const ButtonTemplateCategorySchema = z.enum(["business", "level"]);
export type ButtonTemplateCategory = z.infer<typeof ButtonTemplateCategorySchema>;

export const PanelTemplateCategorySchema = z.enum(["smallPanel", "levelBackground"]);
export type PanelTemplateCategory = z.infer<typeof PanelTemplateCategorySchema>;

export const PatternTemplateCategorySchema = z.enum(["diamond", "coin", "button", "level"]);
export type PatternTemplateCategory = z.infer<typeof PatternTemplateCategorySchema>;

export const BUTTON_TEMPLATE_CATEGORIES: ReadonlyArray<{
  id: ButtonTemplateCategory;
  label: string;
}> = [
  { id: "business", label: "业务类" },
  { id: "level", label: "关卡类" },
];

export const PANEL_TEMPLATE_CATEGORIES: ReadonlyArray<{
  id: PanelTemplateCategory;
  label: string;
}> = [
  { id: "smallPanel", label: "小面板类" },
  { id: "levelBackground", label: "关卡背景类" },
];

export const PATTERN_TEMPLATE_CATEGORIES: ReadonlyArray<{
  id: PatternTemplateCategory;
  label: string;
}> = [
  { id: "diamond", label: "钻石类" },
  { id: "coin", label: "金币类" },
  { id: "button", label: "按钮类" },
  { id: "level", label: "关卡类" },
];

export const DEFAULT_BUTTON_TEMPLATE_CATEGORY: ButtonTemplateCategory = "business";
export const DEFAULT_PANEL_TEMPLATE_CATEGORY: PanelTemplateCategory = "smallPanel";
export const DEFAULT_PATTERN_TEMPLATE_CATEGORY: PatternTemplateCategory = "button";

export const getButtonTemplateCategoryLabel = (category: ButtonTemplateCategory) =>
  BUTTON_TEMPLATE_CATEGORIES.find((entry) => entry.id === category)?.label ?? category;

export const getPanelTemplateCategoryLabel = (category: PanelTemplateCategory) =>
  PANEL_TEMPLATE_CATEGORIES.find((entry) => entry.id === category)?.label ?? category;

export const getPatternTemplateCategoryLabel = (category: PatternTemplateCategory) =>
  PATTERN_TEMPLATE_CATEGORIES.find((entry) => entry.id === category)?.label ?? category;

export const normalizeButtonTemplateCategory = (category: string | undefined): ButtonTemplateCategory =>
  ButtonTemplateCategorySchema.safeParse(category).success
    ? ButtonTemplateCategorySchema.parse(category)
    : DEFAULT_BUTTON_TEMPLATE_CATEGORY;

export const normalizePanelTemplateCategory = (category: string | undefined): PanelTemplateCategory =>
  PanelTemplateCategorySchema.safeParse(category).success
    ? PanelTemplateCategorySchema.parse(category)
    : DEFAULT_PANEL_TEMPLATE_CATEGORY;

export const normalizePatternTemplateCategory = (category: string | undefined): PatternTemplateCategory =>
  PatternTemplateCategorySchema.safeParse(category).success
    ? PatternTemplateCategorySchema.parse(category)
    : DEFAULT_PATTERN_TEMPLATE_CATEGORY;

export const normalizeStretchVisualTemplateCategory = (
  kind: "panel" | "pattern",
  category: string | undefined,
): PanelTemplateCategory | PatternTemplateCategory =>
  kind === "panel"
    ? normalizePanelTemplateCategory(category)
    : normalizePatternTemplateCategory(category);

File: backend/microservice/src/ui/objects/TemplateCategory.scala
package microservice.ui.objects

/** 按钮、面板、图案三类模板的合法 category 枚举与默认值。 */
/** 按钮模板 category 合法值与默认值。 */
object ButtonTemplateCategory {
  val Business: String = "business"
  val Level: String = "level"

  val values: Set[String] = Set(Business, Level)
  val defaultValue: String = Business

  /** 检查 category 字符串是否在合法枚举内。 */
  /** 检查 category 字符串是否在合法枚举内。 */
  /** 检查 category 字符串是否在合法枚举内。 */
  def isValid(value: String): Boolean = values.contains(value)
}

/** 面板拉伸模板 category 合法值与默认值。 */
object PanelTemplateCategory {
  val SmallPanel: String = "smallPanel"
  val LevelBackground: String = "levelBackground"

  val values: Set[String] = Set(SmallPanel, LevelBackground)
  val defaultValue: String = SmallPanel

  def isValid(value: String): Boolean = values.contains(value)
}

/** 图案拉伸模板 category 合法值与默认值。 */
object PatternTemplateCategory {
  val Diamond: String = "diamond"
  val Coin: String = "coin"
  val Button: String = "button"
  val Level: String = "level"

  val values: Set[String] = Set(Diamond, Coin, Button, Level)
  val defaultValue: String = Button

  def isValid(value: String): Boolean = values.contains(value)
}
